server/src/database/commands/clean-inactive-workspaces.command.ts

```
@@ -0,0 +1,237 @@
+interface ActivityReport {
+  displayName: string;
+  maxUpdatedAt: string;
+  inactiveDays: number;
+}
+
+interface SameAsSeedWorkspace {
+  displayName: string;
+}
+
+interface DataCleanResults {
+  activityReport: { [key: string]: ActivityReport };
+  sameAsSeedWorkspaces: { [key: string]: SameAsSeedWorkspace };
+}
+
+@Command({
+  name: 'workspaces:clean-inactive',
+  description: 'Clean inactive workspaces from the public database schema',
+})
+export class DataCleanInactiveCommand extends CommandRunner {
+  constructor(
+    private readonly prismaService: PrismaService,
+    private readonly workspaceService: WorkspaceService,
+    private readonly inquiererService: InquirerService,
+  ) {
+    super();
+  }
+
+  @Option({
+    flags: '-d, --days [inactive days threshold]',
+    description: 'Inactive days threshold',
+    defaultValue: 60,
+  })
+  parseDays(val: string): number {
+    return Number(val);
+  }
+
+  @Option({
+    flags: '-s, --same-as-seed-days [same as seed days threshold]',
+    description: 'Same as seed days threshold',
+    defaultValue: 10,
+  })
+  parseSameAsSeedDays(val: string): number {
+    return Number(val);
+  }
+
+  @Option({
+    flags: '--dry-run [dry run]',
+    description: 'List inactive workspaces without removing them',
+  })
+  parseDryRun(val: string): boolean {
+    return Boolean(val);
+  }
+
+  // We look for public tables which contains workspaceId and updatedAt columns
+  getRelevantTables() {
+    return Object.keys(this.prismaService.client).filter(
+      (name) =>
+        !name.startsWith('_') &&
+        !name.startsWith('$') &&
+        !name.includes('user') &&
+        !name.includes('refreshToken') &&
+        !name.includes('workspace'),
+    );
+  }
+
+  async getTableMaxUpdatedAt(table, workspace) {
+    try {
+      return await this.prismaService.client[table].aggregate({
+        _max: { updatedAt: true },
+        where: { workspaceId: { equals: workspace.id } },
+      });
+    } catch (e) {}
```

why do we try catch and fail silently?

---

in case no updatedAt column exists -> eg: favorites

---

I remove favorites from the tables to check